dali/pipeline/workspace/workspace.h

```diff
@@ -526,20 +527,92 @@ class WorkspaceBase : public ArgumentWorkspace {
   }
 
   /**
- * @brief Returns the index of the sample that this workspace stores
- * in the input/output batch.
- */
+   * @brief Returns the index of the sample that this workspace stores
+   * in the input/output batch.
+   */
   DLL_PUBLIC virtual inline int data_idx() const {
     return 0;
   }
 
   /**
- * @brief Returns the index of the thread that will process this data.
- */
+   * @brief Returns the index of the thread that will process this data.
+   */
   DLL_PUBLIC virtual inline int thread_idx() const {
     return 0;
   }
 
+
+  /// @{
+  /**
+   * Sets the operator ID that this Workspace belongs to.
+   */
+  void SetOperatorId(std::string operator_id) {
+    operator_id_ = std::move(operator_id);
+  }
+
+
+  /**
+   * Returns the operator ID that this Workspace belongs to.
+   */
+  const std::string &GetOperatorId() const {
+    return operator_id_;
+  }
+  /// @}
+
+
+  ///@{
+  /**
+   * Set the whole operator trace map in this workspace.
+   *
+   * Sets the operator trace map that corresponds to all operators in the current iteration.
+   *
+   * Typically, this function shall be called by the Executor, when assigning the Workspace to
+   * the Operator.
+   */
+  void InjectOperatorTraces(std::shared_ptr<operator_trace_map_t> operator_trace_map) {
+    operator_traces_ = std::move(operator_trace_map);
+  }
+
+
+  /**
+   * Set the trace value for the current operator.
+   *
+   * Typically, this function shall be called by an operator in RunImpl or SetupImpl.
+   *
+   * @see operator_trace_map_t
+   */
+  DLL_PUBLIC void SetOperatorTrace(const std::string &trace_key, std::string trace_value) {
+    (*operator_traces_)[GetOperatorId()].insert_or_assign(trace_key, std::move(trace_value));
```

Comment:
Can this potentially modify the trace map from multiple threads? We discussed some scenario close to this problem, assuming that subsequent stages do not run in parallel, it should not happen.

Reply:
Yes, I believe it can (when Operator does it in using ThreadPool). Added a safe lock.

Reply:
On the second thought, I won't.

We decided, that if the operator writes the operator trace in multiple threads - he should handle the thread safety. Furthermore, it's impossible for multiple operators of one iteration to run in parallel, so this also is safe. Lastly, `std::mutex` has no move-ctor, thus implementing the safe lock would require some hacking.